### DISCUSSION TITLE
why does db.go validateOpts() silently adjust incorrect configuration?

### DISCUSSION THREAD

**Discussion by Dieterbe** [prometheus/prometheus#10388] [General]
why does it do this?
why don't we validate and return errors for incorrect configuration? 
now we silently fix things in ways that are quite opaque to the user. they may not get the configuration they expected.

**Comment by roidelapluie** [comment]
Can you please tell us what we are fixing that is "unexpected"? What did you expect?

Most of the options we "correct" would be impossible to deal with, we are not really "changing" anything, mostly setting up defaults.

**Comment by Dieterbe** [comment]
I'm working on some code to support out of order inserts into Head.
this involves setting up some flags in main, passing the options to Head, and of course validating them somewhere.
I would expect there to be 1 central function where the settings can be validated. such function could be called after setting the flags/options to make sure user input is correct, as well as inside NewHead() to doublecheck the options in case the caller is not main().

instead I notice that:
* some options are checked in main() and result in errors if invalid. e.g. `cfg.tsdb.WALSegmentSize`
* other options are checked in validateOpts, and silently adjusted without error.
* other options (e.g. `ChunkRange`) are checked inside NewHead() and result in errors

I looked a bit closer at validateOpts. It checks 6 options. only 2 of them are actually user input(
MaxBlockChunkSegmentSize, MinBlockDuration), the rest is setting defaults, as you say. But: this is very non-obvious. Plus, my original argument remains: if someone passes ` --storage.tsdb.min-block-duration 0` and if this is considered an invalid flag, i would expect an error, not a silent adjustment.

As someone trying to contribute to the code, and trying to follow the best practices, it seems overly complicated to have validation in so many different places, especially to validate user input so late (NewHead) in some cases, and to have a function called validate, which actually doesn't really do what its name suggests.

I would also expect there to be 1 single place where defaults are set (e.g. in main, although this could call a helper function inside the tsdb package for tsdb defaults), followed by letting users override defaults, followed by validation.
Instead, it seems defaults are set in main at `cfg` construction time, in main in the various flag calls, and in validatetopts(), so both prior to having users declare their settings, as well as afterwards. Trying to apply sensible defaults after processing user input, seems rather strange to me.

So TLDR: i think we can simplify and clean up this code, or at least better document validateOpts() to describe what it is for.  This would make the code base more accessible to new developers.

**Reply by roidelapluie** [reply]
Note about `--storage.tsdb.min-block-duration 0`
That flag is hidden and the description says : For use in testing. The flag should not be changed by end users at all.

**Comment by codesome** [comment]
Agree with simplification, and each package having their own ValidateOpts so that the validation is owned by that package. However, I would like to not error, and log the adjustments that we make.
